src/util/auth.rs

```diff
@@ -18,85 +27,225 @@ pub enum AuthenticationError {
+    let req = reqwest::Client::new()
+        .get(dotenvy::var("MINOS_URL").unwrap() + "/user")
+        .header(reqwest::header::USER_AGENT, "Modrinth")
+        .header(reqwest::header::COOKIE, cookies);
+    let res = req.send().await?;
+
+    let res = match res.status() {
+        reqwest::StatusCode::OK => res,
+        reqwest::StatusCode::UNAUTHORIZED => return Err(AuthenticationError::InvalidCredentials),
+        _ => res.error_for_status()?,
+    };
+    Ok(res.json().await?)
+}
+
+pub async fn get_user_from_headers<'a, 'b, E>(
+    headers: &HeaderMap,
     executor: E,
 ) -> Result<User, AuthenticationError>
 where
     E: sqlx::Executor<'a, Database = sqlx::Postgres>,
 {
-    let github_user = get_github_user_from_token(access_token).await?;
-
-    let res = models::User::get_from_github_id(github_user.id, executor).await?;
-
-    match res {
-        Some(result) => Ok(User {
-            id: UserId::from(result.id),
-            github_id: result.github_id.map(|i| i as u64),
-            username: result.username,
-            name: result.name,
-            email: result.email,
-            avatar_url: result.avatar_url,
-            bio: result.bio,
-            created: result.created,
-            role: Role::from_string(&result.role),
-            badges: result.badges,
-            payout_data: Some(UserPayoutData {
-                balance: result.balance,
-                payout_wallet: result.payout_wallet,
-                payout_wallet_type: result.payout_wallet_type,
-                payout_address: result.payout_address,
-            }),
+    let token: Option<&reqwest::header::HeaderValue> = headers.get(AUTHORIZATION);
+    let cookies_unparsed: Option<&reqwest::header::HeaderValue> = headers.get(COOKIE);
+
+    let db_user;
+    let minos_user;
+    // Fetch DB user record and minos user from headers
+    match (token, cookies_unparsed) {
+        (Some(token), _) => {
+            db_user = get_user_record_from_bearer_token(
+                token
+                    .to_str()
+                    .map_err(|_| AuthenticationError::InvalidCredentials)?,
+                executor,
+            )
+            .await?
+            .ok_or_else(|| AuthenticationError::InvalidCredentials)?;
+            minos_user = get_minos_user(
+                &db_user
+                    .kratos_id
+                    .clone()
+                    .ok_or_else(|| AuthenticationError::InvalidCredentials)?,
+            )
+            .await?;
+        }
+        (_, Some(cookies)) => {
+            minos_user = get_minos_user_from_cookies(
+                cookies
+                    .to_str()
+                    .map_err(|_| AuthenticationError::InvalidCredentials)?,
+            )
+            .await?;
+            db_user = models::User::get_from_minos_kratos_id(minos_user.id.clone(), executor)
+                .await?
+                .ok_or_else(|| AuthenticationError::InvalidCredentials)?;
```

Comment:
instead of throwing errors when each auth method fails, shouldn't it try the next one?